crates/bevy_debug_draw/src/debug_draw.rs

```diff
@@ -0,0 +1,200 @@
+use std::f32::consts::TAU;
+
+use bevy_asset::Handle;
+use bevy_ecs::system::Resource;
+use bevy_math::{vec3, Quat, Vec2, Vec3};
+use bevy_render::prelude::{Color, Mesh};
+
+/// A resource with an immediate mode API for drawing lines and wireshapes.
+/// Useful for visual debugging.
+#[derive(Resource)]
+pub struct DebugDraw {
+    positions: Vec<[f32; 3]>,
+    colors: Vec<[f32; 4]>,
+    pub(crate) mesh_handle: Option<Handle<Mesh>>,
+    /// The amount of line segments to use when drawing a circle.
+    ///
+    /// Defaults to `24`.
+    pub circle_segments: u32,
+}
+
+impl Default for DebugDraw {
+    fn default() -> Self {
+        DebugDraw {
+            positions: Vec::new(),
+            colors: Vec::new(),
+            mesh_handle: None,
+            circle_segments: 24,
+        }
+    }
+}
+
+impl DebugDraw {
+    /// Draw a line from `start` to `end`.
+    #[inline]
+    pub fn line(&mut self, start: Vec3, end: Vec3, color: Color) {
+        self.line_gradient(start, end, color, color);
+    }
+
+    /// Draw a line from `start` to `end`.
+    #[inline]
+    pub fn line_gradient(&mut self, start: Vec3, end: Vec3, start_color: Color, end_color: Color) {
+        self.positions.extend([start.to_array(), end.to_array()]);
+        self.colors.extend([
+            start_color.as_linear_rgba_f32(),
+            end_color.as_linear_rgba_f32(),
+        ]);
+    }
+
+    /// Draw a line from `start` to `start + vector`.
+    #[inline]
+    pub fn ray(&mut self, start: Vec3, vector: Vec3, color: Color) {
+        self.ray_gradient(start, vector, color, color);
+    }
+
+    /// Draw a line from `start` to `start + vector`.
+    #[inline]
+    pub fn ray_gradient(
+        &mut self,
+        start: Vec3,
+        vector: Vec3,
+        start_color: Color,
+        end_color: Color,
+    ) {
+        self.line_gradient(start, start + vector, start_color, end_color);
+    }
+
+    /// Draw a circle at `position` with the flat side facing `normal`.
+    #[inline]
+    pub fn circle(&mut self, position: Vec3, normal: Vec3, radius: f32, color: Color) {
+        let rotation = Quat::from_rotation_arc(Vec3::Z, normal);
+        self.positions
+            .extend((0..self.circle_segments).into_iter().flat_map(|i| {
+                let mut angle = i as f32 * TAU / self.circle_segments as f32;
+                let start = rotation * (Vec2::from(angle.sin_cos()) * radius).extend(0.) + position;
+
+                angle += TAU / self.circle_segments as f32;
+                let end = rotation * (Vec2::from(angle.sin_cos()) * radius).extend(0.) + position;
+
+                [start.to_array(), end.to_array()]
+            }));
+
+        self.colors.extend(
+            std::iter::repeat(color.as_linear_rgba_f32()).take(self.circle_segments as usize * 2),
+        );
+    }
+
+    /// Draw a sphere.
+    #[inline]
+    pub fn sphere(&mut self, position: Vec3, radius: f32, color: Color) {
+        self.circle(position, Vec3::X, radius, color);
```

Comment:
Reserving the space in positions and colors here would avoid potential reallocations.